linenoise.hpp

```diff
@@ -145,22 +145,112 @@
 
+/* The linenoiseState structure represents the state during line editing, and
+ * provides methods by which user programs can act on that state. */
+class linenoiseState {
+    public:
+        linenoiseState(const char *prompt = NULL, int stdin_fd = STDIN_FILENO,
+                int stdout_fd = STDOUT_FILENO);
+
+        void EnableMultiLine();
+        void DisableMultiLine();
+
+        bool AddHistory(const char *line);
+        bool LoadHistory(const char *path);
+        bool SaveHistory(const char *path);
+        const std::vector<std::string> &GetHistory() { return history; };
+        bool SetHistoryMaxLen(size_t len);
+
+        bool Readline(std::string &line); // Primary linenoise entry point
+        void RefreshLine(); // Restore current line
+        void WipeLine(); // Temporarily removes line from screen - RefreshLine will restore
+        void ClearScreen(); // Clear terminal window
+
+        /* Register a callback function to be called for tab-completion. */
+        void SetCompletionCallback(CompletionCallback fn) {
+            completionCallback = fn;
+        };
+
+        std::string prompt = std::string("> "); /* Prompt to display. */
+        std::mutex r_mutex;
+
+	// For functional interface
+	void SetMultiLine(bool ml);
+
+    private:
+        std::string Readline(bool &quit);
+        std::string Readline();
+
+        bool linenoiseRaw(std::string &line);
+        int linenoiseEditInsert(const char *cbuf, int clen);
+	void linenoiseEditDelete();
+        void linenoiseEditBackspace();
+        void linenoiseEditDeletePrevWord();
+        int linenoiseEdit();
+	void linenoiseEditHistoryNext(int dir);
+	void linenoiseEditMoveLeft();
+        void linenoiseEditMoveRight();
+        void linenoiseEditMoveHome();
+        void linenoiseEditMoveEnd();
+        void refreshSingleLine();
+        void refreshMultiLine();
+        int completeLine(char *cbuf, int *c);
+
+        CompletionCallback completionCallback;
+
+        int ifd = STDIN_FILENO;          /* Terminal stdin file descriptor. */
+        int ofd = STDOUT_FILENO;         /* Terminal stdout file descriptor. */
+        char *buf = wbuf;                /* Edited line buffer. */
+        int buflen = LINENOISE_MAX_LINE; /* Edited line buffer size. */
+        int pos = 0;                     /* Current cursor position. */
+        int oldcolpos = 0;      /* Previous refresh cursor column position. */
+        int len = 0;            /* Current edited line length. */
+        int lcols = -1;         /* Number of columns in terminal. */
+        int maxrows = 0;        /* Maximum num of rows used so far (multiline mode) */
+        int history_index = -1; /* The history index we are currently editing. */
+        char wbuf[LINENOISE_MAX_LINE] = {'\0'};
+        std::string history_tmpbuf;
+
+        size_t history_max_len = LINENOISE_DEFAULT_HISTORY_MAX_LEN;
+        std::vector<std::string> history;
```

Comment:
Is it possible to suffix all private member variables with an underscore, such as len_, for better code readability in method bodies? Otherwise, it is quite difficult for me to distinguish whether they are member variables or local variables.

Reply:
I added an l_ prefix - will that do the trick?

Reply:
I actually prefer `_` suffix and using the style in my other projects. :)
https://micro-os-plus.github.io/develop/naming-conventions/#private-member-variables-names

Reply:
OK, one second...

Reply:
Switched l_ names to using the _ suffix instead.